pandas/core/frame.py

```diff
@@ -3973,10 +3973,28 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         -------
         dataframe : DataFrame
         """
-        inplace = validate_bool_kwarg(inplace, 'inplace')
+        from pandas import Series
+
         if not isinstance(keys, list):
             keys = [keys]
 
+        # collect elements from "keys" that are not allowed array types
+        col_labels = [x for x in keys
+                      if not isinstance(x, (Series, Index, MultiIndex,
```

Comment:
use the ABC forms here

Reply:
why isn't this just ``isinstance(x, tuple) or is_scalar(x)``?

Reply:
These are the cases (except valid column keys)  that currently *don't* raise a KeyError. I've kept the error reporting exactly the same as before:
```
pd.DataFrame([[0,1], [2,3]]).set_index(map(str,[1,2]))
# KeyError: <map object at 0x000001D380115550>
```

Could be changed, but then I need to know the desired allowed signature / error reporting

Reply:
Tried the ABC Versions, but getting errors that indexes are not hashable. Left it as is for the moment.
```
____________________ TestDataFrameAlterAxes.test_set_index ____________________

self = <pandas.tests.frame.test_alter_axes.TestDataFrameAlterAxes object at 0x0000018FC05ED400>
mixed_frame =                    A         B         C         D  foo
lem22AYh2f -0.599136 -0.811715 -0.135242  0.764100  bar
iBogSI...05  bar
vDdWVTLtoj -0.433490 -1.416721  0.460720 -0.539860  bar
Pw5bUHt4sR  0.180016  1.651358 -1.041539 -0.832112  bar

    def test_set_index(self, mixed_frame):
        df = mixed_frame
        idx = Index(np.arange(len(df))[::-1])

>       df = df.set_index(idx)

pandas\tests\frame\test_alter_axes.py:41:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\core\frame.py:3874: in set_index
    if any(x not in self for x in col_labels):
pandas\core\frame.py:3874: in <genexpr>
    if any(x not in self for x in col_labels):
pandas\core\generic.py:1654: in __contains__
    return key in self._info_axis
pandas\core\indexes\base.py:1994: in __contains__
    hash(key)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Int64Index([29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13,
            12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0],
           dtype='int64')

    def __hash__(self):
>       raise TypeError("unhashable type: %r" % type(self).__name__)
E       TypeError: unhashable type: 'Int64Index'

pandas\core\indexes\base.py:2021: TypeError
_________________ TestDataFrameAlterAxes.test_set_index_cast __________________

self = <pandas.tests.frame.test_alter_axes.TestDataFrameAlterAxes object at 0x0000018FCA8EE668>

    def test_set_index_cast(self):
        # issue casting an index then set_index
        df = DataFrame({'A': [1.1, 2.2, 3.3], 'B': [5.0, 6.1, 7.2]},
                       index=[2010, 2011, 2012])
>       df2 = df.set_index(df.index.astype(np.int32))

pandas\tests\frame\test_alter_axes.py:50:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\core\frame.py:3874: in set_index
    if any(x not in self for x in col_labels):
pandas\core\frame.py:3874: in <genexpr>
    if any(x not in self for x in col_labels):
pandas\core\generic.py:1654: in __contains__
    return key in self._info_axis
pandas\core\indexes\base.py:1994: in __contains__
    hash(key)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Int64Index([2010, 2011, 2012], dtype='int64')

    def __hash__(self):
>       raise TypeError("unhashable type: %r" % type(self).__name__)
E       TypeError: unhashable type: 'Int64Index'

pandas\core\indexes\base.py:2021: TypeError
_______________ TestDataFrameAlterAxes.test_set_index_timezone ________________

self = <pandas.tests.frame.test_alter_axes.TestDataFrameAlterAxes object at 0x0000018FCAD1AEF0>

    def test_set_index_timezone(self):
        # GH 12358
        # tz-aware Series should retain the tz
        idx = to_datetime(["2014-01-01 10:10:10"],
                          utc=True).tz_convert('Europe/Rome')
        df = DataFrame({'A': idx})
>       assert df.set_index(idx).index[0].hour == 11

pandas\tests\frame\test_alter_axes.py:343:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\core\frame.py:3874: in set_index
    if any(x not in self for x in col_labels):
pandas\core\frame.py:3874: in <genexpr>
    if any(x not in self for x in col_labels):
pandas\core\generic.py:1654: in __contains__
    return key in self._info_axis
pandas\core\indexes\base.py:1994: in __contains__
    hash(key)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = DatetimeIndex(['2014-01-01 11:10:10+01:00'], dtype='datetime64[ns, Europe/Rome]', freq=None)

    def __hash__(self):
>       raise TypeError("unhashable type: %r" % type(self).__name__)
E       TypeError: unhashable type: 'DatetimeIndex'

pandas\core\indexes\base.py:2021: TypeError
______________ TestDataFrameAlterAxes.test_dti_set_index_reindex ______________

self = <pandas.tests.frame.test_alter_axes.TestDataFrameAlterAxes object at 0x0000018FCA7FDD68>

    def test_dti_set_index_reindex(self):
        # GH 6631
        df = DataFrame(np.random.random(6))
        idx1 = date_range('2011/01/01', periods=6, freq='M', tz='US/Eastern')
        idx2 = date_range('2013', periods=6, freq='A', tz='Asia/Tokyo')

>       df = df.set_index(idx1)

pandas\tests\frame\test_alter_axes.py:413:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\core\frame.py:3874: in set_index
    if any(x not in self for x in col_labels):
pandas\core\frame.py:3874: in <genexpr>
    if any(x not in self for x in col_labels):
pandas\core\generic.py:1654: in __contains__
    return key in self._info_axis
pandas\core\indexes\base.py:1994: in __contains__
    hash(key)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = DatetimeIndex(['2011-01-31 00:00:00-05:00', '2011-02-28 00:00:00-05:00',
               '2011-03-31 00:00:00-04:00', '... '2011-05-31 00:00:00-04:00', '2011-06-30 00:00:00-04:00'],
              dtype='datetime64[ns, US/Eastern]', freq='M')

    def __hash__(self):
>       raise TypeError("unhashable type: %r" % type(self).__name__)
E       TypeError: unhashable type: 'DatetimeIndex'
```

Reply:
huh? pls use the ABC version, we do this everywhere else. To be very honest I would split this PR up into 2 parts as I think you are actually changing something here but its very very hard to tell.